rohmu/object_storage/google.py

```diff
@@ -703,3 +723,98 @@ def _read_bytes(self, length: int, *, initial_data: Optional[bytes] = None) -> b
+    def next_chunk(self) -> tuple[MediaDownloadProgress, bool]:
+        ...
+
+
+class MediaIoBaseDownloadWithByteRange:
+    """This class is mostly a copy of the googleapiclient's MediaIOBaseDownload class,
+    but with the addition of the support for fetching a specific byte_range.
+
+    """
+
+    def __init__(
+        self,
+        fd: BinaryIO,
+        request: HttpRequest,
+        chunksize: int = DOWNLOAD_CHUNK_SIZE,
+        *,
+        byte_range: tuple[int, int],
+    ) -> None:
+        """Constructor.
+
+        Args:
+          fd: io.Base or file object, The stream in which to write the downloaded
+            bytes.
+          request: googleapiclient.http.HttpRequest, the media request to perform in
+            chunks.
+          chunksize: int, File will be downloaded in chunks of this many bytes.
+          byte_range: Optional[tuple[int, int]], The byterange to fetch
+        """
+        self._fd = fd
+        self._http = request.http
+        self._uri = request.uri
+        self._chunksize = chunksize
+        self._start_range, self._end_range = byte_range
+        self._progress = 0
+        self._range_size = byte_range[1] - byte_range[0] + 1
+        self._done = False
+
+        self._headers = {}
+        for k, v in request.headers.items():
+            # allow users to supply custom headers by setting them on the request
+            # but strip out the ones that are set by default on requests generated by
+            # API methods like Drive's files().get(fileId=...)
+            if k.lower() not in ("accept", "accept-encoding", "user-agent"):
+                self._headers[k] = v
+
+    def next_chunk(self) -> tuple[MediaDownloadProgress, bool]:
+        """Get the next chunk of the download.
+
+        Returns:
+          (status, done): The value of done will be True when the media has been fully
+             downloaded or the total size of the media is unknown.
+
+        Raises:
+          googleapiclient.errors.HttpError if the response was not a 2xx (or a 416 is received and the file is empty)
+          httplib2.HttpLib2Error if a transport error has occurred.
+        """
+        headers = self._headers.copy()
+        chunk_start = self._progress + self._start_range
+        chunk_end = chunk_start + self._chunksize - 1
+        if self._end_range < chunk_end:
+            chunk_end = self._end_range
+        headers["range"] = f"bytes={chunk_start}-{chunk_end}"
+        resp, content = self._http.request(self._uri, "GET", headers=headers)
+
+        total_size: Optional[int] = None
+        if resp.status in (200, 206):
+            if "content-location" in resp and resp["content-location"] != self._uri:
+                self._uri = resp["content-location"]
+            self._progress += len(content)
+            self._fd.write(content)
+
+            if "content-range" in resp:
+                total_size = get_total_size_from_content_range(resp["content-range"])
```

Comment:
is there is a reason content-range would not be in response ?

Reply:
This handles also 200 OKs.

From the specs it is valid for a server to return a 200 OK with the whole object if it makes sense to do so/the server/a proxy in between does not support Range (e.g. I ask range `0-100` for some content and the content is 10 bytes, the server can either return a 206 with `Content-Range 0-10/10` or a `200 OK` with `Content-Length: 10` and no `Content-Range` required from what I understand).

This handling of 200/206 and these headers is taken from the library so it should be correct or we need to bugfix the official implementation too.